### PR TITLE
Feature layer time offset

### PR DESCRIPTION
**Pull request by nCastle1**
Note: this pull request is not yet ready for review. As written, the general concept is demonstrated with a different type of layer. Once the runtime supports time functionality for Feature Layer, the sample will be updated and a review will be requested.

### CODE REVIEW

**Comment by nCastle1** [comment]
@dkemlage Can you please give this a look when you get a chance?

File: src/WPF/ArcGISRuntime.WPF.Samples/Samples/MapView/FeatureLayerTimeOffset/FeatureLayerTimeOffset.xaml.cs [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+
+namespace ArcGISRuntime.WPF.Samples.FeatureLayerTimeOffset
+{
+    public partial class FeatureLayerTimeOffset
+    {
+        // Hold the feature layer URI
+        private Uri _featureLayerUri = new Uri("https://sampleserver6.arcgisonline.com/arcgis/rest/services/Hurricanes/MapServer/0");
+
+        // Hold a reference to the original time extent
+        private TimeExtent originalExtent;
```
Comment (dkemlage) [inline src/WPF/ArcGISRuntime.WPF.Samples/Samples/MapView/FeatureLayerTimeOffset/FeatureLayerTimeOffset.xaml.cs]: Global scope variable should be with an underscore ( _ ). Ex: _originalExtent.

File: src/WPF/ArcGISRuntime.WPF.Samples/Samples/MapView/FeatureLayerTimeOffset/FeatureLayerTimeOffset.xaml.cs [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+        private async void Initialize()
+        {
+            // Create new Map
+            Map myMap = new Map(Basemap.CreateOceans());
+
+            // Create the hurricanes feature layer once
+            FeatureLayer noOffsetLayer = new FeatureLayer(_featureLayerUri);
+
+            // Apply a blue dot renderer to distinguish hurricanes without offsets
+            noOffsetLayer.Renderer = new SimpleRenderer(new SimpleMarkerSymbol(SimpleMarkerSymbolStyle.Circle, System.Windows.Media.Colors.Blue, 10));
```
Comment (dkemlage) [inline src/WPF/ArcGISRuntime.WPF.Samples/Samples/MapView/FeatureLayerTimeOffset/FeatureLayerTimeOffset.xaml.cs]: The code would read cleaner, if this line were broken into 2 lines of code. One line of code for the creation of the SimpleMarkerSymbol and another to create the SimpleRenderer. Chaining of code together makes it harder to understand what is occurring and requires a lot of scrolling to read what is going on.

File: src/WPF/ArcGISRuntime.WPF.Samples/Samples/MapView/FeatureLayerTimeOffset/FeatureLayerTimeOffset.xaml.cs [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+
+            // Add the non-offset layer to the map
+            myMap.OperationalLayers.Add(noOffsetLayer);
+
+            // Create the offset hurricanes feature layer
+            FeatureLayer withOffsetLayer = new FeatureLayer(_featureLayerUri);
+
+            // Apply a red dot renderer to distinguish these hurricanes from the non-offset hurricanes
+            withOffsetLayer.Renderer = new SimpleRenderer(new SimpleMarkerSymbol(SimpleMarkerSymbolStyle.Circle, System.Windows.Media.Colors.Red, 10));
```
Comment (dkemlage) [inline src/WPF/ArcGISRuntime.WPF.Samples/Samples/MapView/FeatureLayerTimeOffset/FeatureLayerTimeOffset.xaml.cs]: The code would read cleaner, if this line were broken into 2 lines of code. One line of code for the creation of the SimpleMarkerSymbol and another to create the SimpleRenderer. Chaining of code together makes it harder to understand what is occurring and requires a lot of scrolling to read what is going on.

File: src/WPF/ArcGISRuntime.WPF.Samples/Samples/MapView/FeatureLayerTimeOffset/FeatureLayerTimeOffset.xaml.cs [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+
+            // Apply the time offset (red hurricane dots will be from 10 days before the current extent)
+            withOffsetLayer.TimeOffset = new TimeValue(10, Esri.ArcGISRuntime.ArcGISServices.TimeUnit.Days);
```
Comment (dkemlage) [inline src/WPF/ArcGISRuntime.WPF.Samples/Samples/MapView/FeatureLayerTimeOffset/FeatureLayerTimeOffset.xaml.cs]: The code would read cleaner, if this line were broken into 2 lines of code. One line of code for the creation of the TimeValue and another to set the FeatureLayer.TimeOffset property. Chaining of code together makes it harder to understand what is occurring and requires a lot of scrolling to read what is going on.

File: src/WPF/ArcGISRuntime.WPF.Samples/Samples/MapView/FeatureLayerTimeOffset/FeatureLayerTimeOffset.xaml.cs [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+
+        private void UpdateTimeExtent()
+        {
+            // Get the value of the slider
+            double value = MyTimeSlider.Value;
+
+            // Calculate the number of days that value corresponds to
+            // 1. Get the interval
+            TimeSpan interval = originalExtent.EndTime - originalExtent.StartTime;
+            // 2. Store the interval as days
```
Comment (dkemlage) [inline src/WPF/ArcGISRuntime.WPF.Samples/Samples/MapView/FeatureLayerTimeOffset/FeatureLayerTimeOffset.xaml.cs]: Add a line break for readability.

File: src/WPF/ArcGISRuntime.WPF.Samples/Samples/MapView/FeatureLayerTimeOffset/FeatureLayerTimeOffset.xaml.cs [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+            double days = interval.TotalDays;
+            // 3. Scale the interval by the value from the slider
```
Comment (dkemlage) [inline src/WPF/ArcGISRuntime.WPF.Samples/Samples/MapView/FeatureLayerTimeOffset/FeatureLayerTimeOffset.xaml.cs]: Add a line break for readability.

File: src/WPF/ArcGISRuntime.WPF.Samples/Samples/MapView/FeatureLayerTimeOffset/FeatureLayerTimeOffset.xaml.cs [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
+            double desiredInterval = value * days;
+            // 4. Create a new TimeSpan
```
Comment (dkemlage) [inline src/WPF/ArcGISRuntime.WPF.Samples/Samples/MapView/FeatureLayerTimeOffset/FeatureLayerTimeOffset.xaml.cs]: Add a line break for readability.

File: src/Forms/Shared/Samples/MapView/FeatureLayerTimeOffset/FeatureLayerTimeOffset.xaml.cs [hunk 8]
```diff
@@ -0,0 +1,116 @@
...
+// Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the specific
+// language governing permissions and limitations under the License.
+
+using Esri.ArcGISRuntime;
+using Esri.ArcGISRuntime.Mapping;
+using Esri.ArcGISRuntime.Symbology;
+using System;
+using Xamarin.Forms;
+
```
Comment (dkemlage) [inline src/Forms/Shared/Samples/MapView/FeatureLayerTimeOffset/FeatureLayerTimeOffset.xaml.cs]: Add the following compiler conditional statements to overcome platform specific use of colors:
```
#if WINDOWS_UWP
using Colors = Windows.UI.Colors;
#else
using Colors = System.Drawing.Color;
#endif
```

File: src/Forms/Shared/Samples/MapView/FeatureLayerTimeOffset/FeatureLayerTimeOffset.xaml.cs [hunk 9]
```diff
@@ -0,0 +1,116 @@
...
+        private async void Initialize()
+        {
+            // Create new Map
+            Map myMap = new Map(Basemap.CreateOceans());
+
+            // Create the hurricanes feature layer once
+            FeatureLayer noOffsetLayer = new FeatureLayer(_featureLayerUri);
+
+            // Apply a blue dot renderer to distinguish hurricanes without offsets
+            noOffsetLayer.Renderer = new SimpleRenderer(new SimpleMarkerSymbol(SimpleMarkerSymbolStyle.Circle, Windows.UI.Colors.Blue, 10));
```
Comment (dkemlage) [inline src/Forms/Shared/Samples/MapView/FeatureLayerTimeOffset/FeatureLayerTimeOffset.xaml.cs]: Modify the following line of code to make use of platform specific color issues:
```
            noOffsetLayer.Renderer = new SimpleRenderer(new SimpleMarkerSymbol(SimpleMarkerSymbolStyle.Circle, Colors.Blue, 10));

```

File: src/Forms/Shared/Samples/MapView/FeatureLayerTimeOffset/FeatureLayerTimeOffset.xaml.cs [hunk 10]
```diff
@@ -0,0 +1,116 @@
...
+
+            // Add the non-offset layer to the map
+            myMap.OperationalLayers.Add(noOffsetLayer);
+
+            // Create the offset hurricanes feature layer
+            FeatureLayer withOffsetLayer = new FeatureLayer(_featureLayerUri);
+
+            // Apply a red dot renderer to distinguish these hurricanes from the non-offset hurricanes
+            withOffsetLayer.Renderer = new SimpleRenderer(new SimpleMarkerSymbol(SimpleMarkerSymbolStyle.Circle, Windows.UI.Colors.Red, 10));
```
Comment (dkemlage) [inline src/Forms/Shared/Samples/MapView/FeatureLayerTimeOffset/FeatureLayerTimeOffset.xaml.cs]: Modify the following line of code to make use of platform specific color issues:
```
            withOffsetLayer.Renderer = new SimpleRenderer(new SimpleMarkerSymbol(SimpleMarkerSymbolStyle.Circle, Colors.Red, 10));
```

**Comment by nCastle1** [comment]
@dkemlage My apologies, I had forgotten to push the commit I made 5 days ago before I requested your review. As a result, some of the changes you suggested had already been made.

I've gone back and made the other changes you suggested that. Can you please take another look?

Thanks! :)

**Comment by nCastle1** [comment]
@ThadT, Don and I are having some issues reproducing a UWP build problem. Can you please take a look and see if you can build the UWP project?

**Comment by ThadT** [comment]
@nCastle1 @dkemlage 
I was able to build and run the UWP version of the sample without issues.

**Comment by dkemlage** [comment]
@nCastle1 

I retested all of the samples and UWP and Xamarin.Forms.UWP are working great. 

I am still seeing issues with Xamarin.Forms.Android and Xamarin.Android not displaying any red and blue points. After some additional testing, it seems that several other FeatureLayer based samples are no longer showing point features. This could be a regression that needs to be fixed. 

So I am OK with merging the new FeatureLayerTimeOffset sample in the code base.

Thanks,
-don

**Pull request merged**